OpenRA.Game/Network/Connection.cs

```diff
@@ -304,16 +304,22 @@ void IConnection.Receive(OrderManager orderManager)
 					orderManager.ReceiveDisconnect(disconnect.ClientId, disconnect.Frame);
 				else if (OrderIO.TryParseSync(p.Data, out var sync))
 					orderManager.ReceiveSync(sync);
-				else if (OrderIO.TryParseAck(p, out var ackFrame))
+				else if (OrderIO.TryParseAck(p, out var ackFrame, out var count))
 				{
-					if (!sentOrders.TryDequeue(out var q))
-						throw new InvalidOperationException("Received Ack with empty send queue");
+					var orders = new List<OrderPacket>();
```

Comment:
We can avoid allocating and copying things twice in the common case by doing something like:

```csharp
if (count > sentOrders.Count)
	throw new InvalidOperationException($"Received Ack for {count} > {sentOrders.Count} frames.");

OrderPacket packet;
if (count != 1)
{
	var orders = Enumerable.Range(0, count)
		.Select(i => sentOrders.Dequeue().Orders);
	packet = OrderPacket.Combine(orders);
}
else
	packet = sentOrders.Dequeue().Orders;
```